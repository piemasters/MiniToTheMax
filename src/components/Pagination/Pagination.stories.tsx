import type { Meta, StoryObj } from '@storybook/react-vite';

import { Pagination } from './Pagination';

const meta: Meta<typeof Pagination> = {
  component: Pagination,
  argTypes: {
    numPages: {
      name: 'numPages',
      description: 'The number of pages to display between Previous and Next',
      defaultValue: 20,
      control: { type: 'number' },
    },
    currentPage: {
      name: 'currentPage',
      description: 'The current page number',
      defaultValue: 3,
      control: { type: 'number' },
    },
    maxPages: {
      name: 'maxPages',
      description: 'The maximum number of pages to show at once',
      defaultValue: 10,
      control: { type: 'number' },
    },
    baseUrl: {
      name: 'baseUrl',
      description:
        'The base URL the prefixes each of the generated page strings/numbers',
      defaultValue: '/blog/',
      control: { type: 'text' },
    },
  },
};

export default meta;
type Story = StoryObj<typeof Pagination>;

export const Primary: Story = {
  args: {
    numPages: 20,
    currentPage: 3,
    maxPages: 10,
    baseUrl: '/',
  },
};
