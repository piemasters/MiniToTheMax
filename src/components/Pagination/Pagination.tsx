import type { FC } from 'react';
import { css, useTheme } from '@emotion/react';

import { PageLink } from '../PageLink';
import { Theme } from '../../styles/theme';
import { getPages } from '../../util';

export interface PaginationProps {
  isFirst: boolean;
  isLast: boolean;
  prevPage: string;
  nextPage: string;
  numPages: number;
  currentPage: number;
  maxPages?: number;
  baseUrl: string;
}

export const Pagination: FC<PaginationProps> = ({
  numPages,
  currentPage,
  maxPages = 8,
  baseUrl,
}) => {
  const theme = useTheme() as Theme;

  const containerStyle = css`
    align-items: center;
    display: flex;
    flex-wrap: wrap;
    justify-content: space-between;
    list-style: none;
    padding: 0;
    margin: 2rem 0;
  `;

  const linkStyle = (i: number) => {
    return css`
      background: ${i === currentPage ? theme.colors.hyperlinkActive : ''};
      color: ${i === currentPage ? '#ffffff' : theme.colors.hyperlink};
      padding: 0.5rem;
      text-decoration: none;
      &:visited {
        color: ${i === currentPage ? '#ffffff' : theme.colors.hyperlink};
      }
      &:hover {
        color: ${i === currentPage ? '#ffffff' : theme.colors.hyperlinkActive};
      }
    `;
  };

  // Return an array of pages to repeat
  const pages = getPages({ numPages, maxPages, currentPage });

  return (
    <ul css={containerStyle} data-testid="pagination">
      <li>
        <PageLink
          to={`${baseUrl}`}
          type={'cover'}
          direction={'right'}
          disabled={currentPage === 1}
        >
          &lt;&lt;
        </PageLink>
      </li>
      {/*<li>*/}
      {/*  <PageLink*/}
      {/*    to={`${baseUrl}${currentPage - 1 === 1 ? '' : currentPage - 1}`}*/}
      {/*    type={'cover'}*/}
      {/*    direction={'right'}*/}
      {/*    disabled={currentPage === 1}*/}
      {/*  >*/}
      {/*    &lt;*/}
      {/*  </PageLink>*/}
      {/*</li>*/}
      {pages.map((page: number, index: number) => (
        <li key={index}>
          <PageLink
            to={`${baseUrl}${page === 1 ? '' : page}`}
            type={'cover'}
            direction={page > currentPage ? 'left' : 'right'}
            linkStyle={linkStyle(page)}
          >
            {page}
          </PageLink>
        </li>
      ))}
      {/*<li>*/}
      {/*  <PageLink*/}
      {/*    to={`${baseUrl}${currentPage + 1}`}*/}
      {/*    type={'cover'}*/}
      {/*    direction={'right'}*/}
      {/*    disabled={currentPage === numPages}*/}
      {/*  >*/}
      {/*    &gt;*/}
      {/*  </PageLink>*/}
      {/*</li>*/}
      <li>
        <PageLink
          to={`${baseUrl}${numPages}`}
          type={'cover'}
          direction={'right'}
          disabled={currentPage === numPages}
        >
          &gt;&gt;
        </PageLink>
      </li>
    </ul>
  );
};

export default Pagination;
