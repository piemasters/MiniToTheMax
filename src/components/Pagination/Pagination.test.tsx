import { getByTestId, render } from '@testing-library/react';

import { Pagination } from './Pagination';

describe('Pagination', () => {
  test('renders correctly', () => {
    const { container } = render(
      <Pagination numPages={5} currentPage={3} maxPages={10} baseUrl={'/'} />
    );

    expect(
      getByTestId(container as HTMLElement, 'pagination')
    ).toHaveTextContent('<<12345>>');
    expect(container).toMatchSnapshot();
  });
});
