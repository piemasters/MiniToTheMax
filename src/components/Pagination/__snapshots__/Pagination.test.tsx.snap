// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders correctly 1`] = `
<div>
  <ul
    class="flex flex-wrap justify-between items-center p-0 my-8 list-none"
    data-testid="pagination"
  >
    <li>
      <a
        aria-current="page"
        class="text-blue-500 visited:text-blue-500 hover:text-blue-700 no-underline"
        data-testid="page-link"
        href="/"
      >
        &lt;&lt;
      </a>
    </li>
    <li>
      <a
        aria-current="page"
        class="p-2 no-underline text-blue-500 hover:bg-blue-500 hover:text-white "
        data-testid="page-link"
        href="/"
      >
        1
      </a>
    </li>
    <li>
      <a
        class="p-2 no-underline text-blue-500 hover:bg-blue-500 hover:text-white "
        data-testid="page-link"
        href="/2"
      >
        2
      </a>
    </li>
    <li>
      <a
        class="p-2 no-underline text-blue-500 hover:bg-blue-500 hover:text-white bg-blue-500 text-white"
        data-testid="page-link"
        href="/3"
      >
        3
      </a>
    </li>
    <li>
      <a
        class="p-2 no-underline text-blue-500 hover:bg-blue-500 hover:text-white "
        data-testid="page-link"
        href="/4"
      >
        4
      </a>
    </li>
    <li>
      <a
        class="p-2 no-underline text-blue-500 hover:bg-blue-500 hover:text-white "
        data-testid="page-link"
        href="/5"
      >
        5
      </a>
    </li>
    <li>
      <a
        class="text-blue-500 visited:text-blue-500 hover:text-blue-700 no-underline"
        data-testid="page-link"
        href="/5"
      >
        &gt;&gt;
      </a>
    </li>
  </ul>
</div>
`;
